src/mesh/amr_loadbalance.cpp

```diff
@@ -45,6 +45,241 @@
+                         Variable<Real> *var_in, Variable<Real> *var, MeshBlock *pmb,
+                         Mesh *pmesh) {
+  // TODO(LFR): Set the index ranges appropriately for general topological element
+  // variables
+  static const IndexRange ib = pmb->c_cellbounds.GetBoundsI(IndexDomain::interior);
+  static const IndexRange jb = pmb->c_cellbounds.GetBoundsJ(IndexDomain::interior);
+  static const IndexRange kb = pmb->c_cellbounds.GetBoundsK(IndexDomain::interior);
+
+  const int ox1 = ((fine_loc.lx1 & 1LL) == 1LL);
+  const int ox2 = ((fine_loc.lx2 & 1LL) == 1LL);
+  const int ox3 = ((fine_loc.lx3 & 1LL) == 1LL);
+
+  const int ks = (ox3 == 0) ? 0 : (kb.e - kb.s + 1);
+  const int js = (ox2 == 0) ? 0 : (jb.e - jb.s + 1);
+  const int is = (ox1 == 0) ? 0 : (ib.e - ib.s + 1);
+
+  int test = 1;
+#ifdef MPI_PARALLEL
+  MPI_Comm comm = pmesh->GetMPIComm(var->label());
+  int tag = CreateAMRMPITag(lid_recv, ox1, ox2, ox3);
+  MPI_Status status;
+  PARTHENON_MPI_CHECK(MPI_Iprobe(send_rank, tag, comm, &test, &status));
+#endif
+  if (test) {
+    int size = var_in->IsAllocated();
+#ifdef MPI_PARALLEL
+    PARTHENON_MPI_CHECK(MPI_Get_count(&status, MPI_PARTHENON_REAL, &size));
+#endif
+    if (size > 0) {
+      if (!pmb->IsAllocated(var->label())) pmb->AllocateSparse(var->label());
+      auto cb = var_in->coarse_s;
+#ifdef MPI_PARALLEL
+      // This has to be an MPI_Recv w/o buffering
+      PARTHENON_MPI_CHECK(MPI_Recv(var->coarse_s.data(), var->coarse_s.size(),
+                                   MPI_PARTHENON_REAL, send_rank, tag, comm,
+                                   MPI_STATUS_IGNORE));
+      cb = var->coarse_s;
+#endif
+      auto fb = var->data;
+      const int nt = fb.GetDim(6) - 1;
+      const int nu = fb.GetDim(5) - 1;
+      const int nv = fb.GetDim(4) - 1;
+      parthenon::par_for(
+          DEFAULT_LOOP_PATTERN, "FillSameRankCoarseToFineAMR", DevExecSpace(), 0, nt, 0,
+          nu, 0, nv, kb.s, kb.e, jb.s, jb.e, ib.s, ib.e,
+          KOKKOS_LAMBDA(const int t, const int u, const int v, const int k, const int j,
+                        const int i) {
+            fb(t, u, v, k + ks, j + js, i + is) = cb(t, u, v, k, j, i);
+          });
+      // We have to block here w/o buffering so that the write is guaranteed to be
+      // finished before we get here again
+      Kokkos::fence();
+    } else {
+#ifdef MPI_PARALLEL
+      PARTHENON_MPI_CHECK(MPI_Recv(var->data.data(), 0, MPI_PARTHENON_REAL, send_rank,
+                                   tag, comm, MPI_STATUS_IGNORE));
+#endif
+    }
+  }
+
+  return test;
+}
+
+#ifdef MPI_PARALLEL
+MPI_Request SendSameToSame(int lid_recv, int dest_rank, Variable<Real> *var,
+                           MeshBlock *pmb, Mesh *pmesh) {
+  MPI_Request req;
+  MPI_Comm comm = pmesh->GetMPIComm(var->label());
+  int tag = CreateAMRMPITag(lid_recv, 0, 0, 0);
+  if (var->IsAllocated()) {
+    // This assumes we have two ghost zones
```

Comment:
Does this imply that running with `nghost != 2` does not work correctly any more?
If so, I'd consider this blocking (though the fix should be straightforward, isn't it)?

Reply:
Also could you please add a comment above those lines for what they're doing.
"Update (de)reference counter" (or similar -- if I follow correctly)

Reply:
This should read "at *least* two ghost zones", sorry. I have added a comment to the code explaining what is going on and added a `PARTHENON_REQUIRE(nghost>1)`. Let me know if you still think this is blocking.

Reply:
Perfect.